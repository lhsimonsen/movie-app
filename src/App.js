import React, { Component } from 'react';
import styled from 'react-emotion';
import axios from 'axios';
import {flattenMovie, sortByProp, filterByValue} from './utils';
import {colors, borders, breakpoints, endpoint} from './utils/constants';
import MovieList from './components/MovieList';
import MovieDetail from './components/MovieDetail';
import SortBy from './components/SortBy';
import SearchBar from './components/SearchBar';

const R = require('ramda');

const Wrapper = styled.div`
  font-family: sans-serif;
  color: ${colors.dark};
  display: flex;
  flex-direction: column;
`;

const Outer = styled.div`
  display: block;

  @media (min-width: ${breakpoints.tablet}px) {
    flex: 1;
    display: flex;
    flex-direction: row;
  }

  .dark {
    background: ${colors.primary};
    border: ${borders.primary};
    padding: 15px 10px;
  }
`;

const Inner = styled.div`
  position: relative;
  flex: 1;

  .base {
    display: flex;
  }
  .left {
    border-right: ${borders.primary};
    display: block;

    @media (min-width: ${breakpoints.tablet}px) {
      display: flex;
    }
  }
  .right {
    padding: 30px;
    display: block;

    @media (min-width: ${breakpoints.tablet}px) {
      display: flex;
    }
  }
`;

class App extends Component {
  constructor() {
    super();

    this.state = {
      currentMovie: null,
      sortBy: null,
      searchValue: "",
      movieData: []
    }

    this.getSortedMovies = this.getSortedMovies.bind(this);
    this.getFilteredMovies = this.getFilteredMovies.bind(this);
    this.setSortBy = this.setSortBy.bind(this);
  }

  componentDidMount() {
    axios.get(endpoint)
      .then(res => this.setState({movieData: res.data}))
      .catch(err => console.log(err.message));
  }

  getNormalizedData(data) {
    return R.map(flattenMovie, data);
  }

  getSelectedMovie(id, movies) {
    return R.find(R.propEq('id', id), movies);
  }

  getSortedMovies(movies) {
    const {sortBy} = this.state;
    return R.isNil(sortBy) ? movies : sortByProp(sortBy)(movies);
  }

  getFilteredMovies(movies) {
    const {searchValue} = this.state;
    return filterByValue(searchValue)(movies);
  }

  setSortBy(value) {
    this.setState({sortBy: value});
  }

  setSearchValue(value) {
    this.setState({searchValue: value});
  }

  render() {
    const {currentMovie, searchValue} = this.state;
    const movies = R.compose(this.getNormalizedData)(this.state.movieData);
    const filteredMovies = R.compose(this.getFilteredMovies, this.getSortedMovies)(movies);

    return (
      <Wrapper>
        <Outer className="dark">
          <Inner className="base">
            <SortBy setSortBy={value => this.setSortBy(value)}/>
            <SearchBar
              searchValue={searchValue}
              setSearchValue={val => this.setSearchValue(val)}
            />
          </Inner>
        </Outer>
        <Outer>
          <Inner className="left">
            <MovieList
              movies={filteredMovies}
              onClick={id => this.setState({currentMovie: id})}
            />
          </Inner>
          <Inner className="right">
            <MovieDetail movie={this.getSelectedMovie(currentMovie, movies)} />
          </Inner>
        </Outer>
      </Wrapper>
    );
  }
}

export default App;
