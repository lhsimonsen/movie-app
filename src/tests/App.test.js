import React from 'react';
import {shallow} from 'enzyme';
import App from '../App';

const flushPromises = () => new Promise(resolve => setImmediate(resolve));

it('App snapshot test', async () => {
  const wrapper = shallow(<App />);
  await flushPromises();
  wrapper.update();
  expect(wrapper).toMatchSnapshot();
});

it('renders a movie list', () => {
  const component = shallow(<App />);
  expect(component.find("MovieList").length).toEqual(1);
});

it('renders a movie detail', () => {
  const component = shallow(<App />);
  expect(component.find("MovieDetail").length).toEqual(1);
});

it('renders a sort by button', () => {
  const component = shallow(<App />);
  expect(component.find("SortBy").length).toEqual(1);
});

it('renders a search bar', () => {
  const component = shallow(<App />);
  expect(component.find("SearchBar").length).toEqual(1);
});
