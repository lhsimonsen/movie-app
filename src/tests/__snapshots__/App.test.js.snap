// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot test 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <styled(Component)
          className="dark"
>
          <styled(Component)
                    className="base"
          >
                    <SortBy
                              setSortBy={[Function]}
                    />
                    <SearchBar
                              searchValue=""
                              setSearchValue={[Function]}
                    />
          </styled(Component)>
</styled(Component)>,
        <styled(Component)>
          <styled(Component)
                    className="left"
          >
                    <MovieList
                              movies={
                                        Array [
                                                  Object {
                                                    "director": "Foo Bar",
                                                    "episode_id": 22,
                                                    "id": 1,
                                                    "opening_crawl": "foo foo foo",
                                                    "release_date": "2001-01-01",
                                                    "title": "Foo",
                                                  },
                                                  Object {
                                                    "director": "Bar Baz",
                                                    "episode_id": 13,
                                                    "id": 2,
                                                    "opening_crawl": "bar bar bar",
                                                    "release_date": "2001-02-02",
                                                    "title": "Bar",
                                                  },
                                                  Object {
                                                    "director": "Baz Biz",
                                                    "episode_id": 37,
                                                    "id": 3,
                                                    "opening_crawl": "baz baz baz",
                                                    "release_date": "2003-03-03",
                                                    "title": "Baz",
                                                  },
                                                ]
                              }
                              onClick={[Function]}
                    />
          </styled(Component)>
          <styled(Component)
                    className="right"
          >
                    <MovieDetail
                              movie={undefined}
                    />
          </styled(Component)>
</styled(Component)>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <styled(Component)
            className="base"
>
            <SortBy
                        setSortBy={[Function]}
            />
            <SearchBar
                        searchValue=""
                        setSearchValue={[Function]}
            />
</styled(Component)>,
          "className": "dark",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <SortBy
                setSortBy={[Function]}
/>,
              <SearchBar
                searchValue=""
                setSearchValue={[Function]}
/>,
            ],
            "className": "base",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "setSortBy": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "searchValue": "",
                "setSearchValue": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <styled(Component)
              className="left"
>
              <MovieList
                            movies={
                                          Array [
                                                        Object {
                                                          "director": "Foo Bar",
                                                          "episode_id": 22,
                                                          "id": 1,
                                                          "opening_crawl": "foo foo foo",
                                                          "release_date": "2001-01-01",
                                                          "title": "Foo",
                                                        },
                                                        Object {
                                                          "director": "Bar Baz",
                                                          "episode_id": 13,
                                                          "id": 2,
                                                          "opening_crawl": "bar bar bar",
                                                          "release_date": "2001-02-02",
                                                          "title": "Bar",
                                                        },
                                                        Object {
                                                          "director": "Baz Biz",
                                                          "episode_id": 37,
                                                          "id": 3,
                                                          "opening_crawl": "baz baz baz",
                                                          "release_date": "2003-03-03",
                                                          "title": "Baz",
                                                        },
                                                      ]
                            }
                            onClick={[Function]}
              />
</styled(Component)>,
            <styled(Component)
              className="right"
>
              <MovieDetail
                            movie={undefined}
              />
</styled(Component)>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <MovieList
                movies={
                                Array [
                                                Object {
                                                  "director": "Foo Bar",
                                                  "episode_id": 22,
                                                  "id": 1,
                                                  "opening_crawl": "foo foo foo",
                                                  "release_date": "2001-01-01",
                                                  "title": "Foo",
                                                },
                                                Object {
                                                  "director": "Bar Baz",
                                                  "episode_id": 13,
                                                  "id": 2,
                                                  "opening_crawl": "bar bar bar",
                                                  "release_date": "2001-02-02",
                                                  "title": "Bar",
                                                },
                                                Object {
                                                  "director": "Baz Biz",
                                                  "episode_id": 37,
                                                  "id": 3,
                                                  "opening_crawl": "baz baz baz",
                                                  "release_date": "2003-03-03",
                                                  "title": "Baz",
                                                },
                                              ]
                }
                onClick={[Function]}
/>,
              "className": "left",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "movies": Array [
                  Object {
                    "director": "Foo Bar",
                    "episode_id": 22,
                    "id": 1,
                    "opening_crawl": "foo foo foo",
                    "release_date": "2001-01-01",
                    "title": "Foo",
                  },
                  Object {
                    "director": "Bar Baz",
                    "episode_id": 13,
                    "id": 2,
                    "opening_crawl": "bar bar bar",
                    "release_date": "2001-02-02",
                    "title": "Bar",
                  },
                  Object {
                    "director": "Baz Biz",
                    "episode_id": 37,
                    "id": 3,
                    "opening_crawl": "baz baz baz",
                    "release_date": "2003-03-03",
                    "title": "Baz",
                  },
                ],
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <MovieDetail
                movie={undefined}
/>,
              "className": "right",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "movie": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <styled(Component)
            className="dark"
>
            <styled(Component)
                        className="base"
            >
                        <SortBy
                                    setSortBy={[Function]}
                        />
                        <SearchBar
                                    searchValue=""
                                    setSearchValue={[Function]}
                        />
            </styled(Component)>
</styled(Component)>,
          <styled(Component)>
            <styled(Component)
                        className="left"
            >
                        <MovieList
                                    movies={
                                                Array [
                                                            Object {
                                                              "director": "Foo Bar",
                                                              "episode_id": 22,
                                                              "id": 1,
                                                              "opening_crawl": "foo foo foo",
                                                              "release_date": "2001-01-01",
                                                              "title": "Foo",
                                                            },
                                                            Object {
                                                              "director": "Bar Baz",
                                                              "episode_id": 13,
                                                              "id": 2,
                                                              "opening_crawl": "bar bar bar",
                                                              "release_date": "2001-02-02",
                                                              "title": "Bar",
                                                            },
                                                            Object {
                                                              "director": "Baz Biz",
                                                              "episode_id": 37,
                                                              "id": 3,
                                                              "opening_crawl": "baz baz baz",
                                                              "release_date": "2003-03-03",
                                                              "title": "Baz",
                                                            },
                                                          ]
                                    }
                                    onClick={[Function]}
                        />
            </styled(Component)>
            <styled(Component)
                        className="right"
            >
                        <MovieDetail
                                    movie={undefined}
                        />
            </styled(Component)>
</styled(Component)>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <styled(Component)
              className="base"
>
              <SortBy
                            setSortBy={[Function]}
              />
              <SearchBar
                            searchValue=""
                            setSearchValue={[Function]}
              />
</styled(Component)>,
            "className": "dark",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <SortBy
                  setSortBy={[Function]}
/>,
                <SearchBar
                  searchValue=""
                  setSearchValue={[Function]}
/>,
              ],
              "className": "base",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "setSortBy": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "searchValue": "",
                  "setSearchValue": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <styled(Component)
                className="left"
>
                <MovieList
                                movies={
                                                Array [
                                                                Object {
                                                                  "director": "Foo Bar",
                                                                  "episode_id": 22,
                                                                  "id": 1,
                                                                  "opening_crawl": "foo foo foo",
                                                                  "release_date": "2001-01-01",
                                                                  "title": "Foo",
                                                                },
                                                                Object {
                                                                  "director": "Bar Baz",
                                                                  "episode_id": 13,
                                                                  "id": 2,
                                                                  "opening_crawl": "bar bar bar",
                                                                  "release_date": "2001-02-02",
                                                                  "title": "Bar",
                                                                },
                                                                Object {
                                                                  "director": "Baz Biz",
                                                                  "episode_id": 37,
                                                                  "id": 3,
                                                                  "opening_crawl": "baz baz baz",
                                                                  "release_date": "2003-03-03",
                                                                  "title": "Baz",
                                                                },
                                                              ]
                                }
                                onClick={[Function]}
                />
</styled(Component)>,
              <styled(Component)
                className="right"
>
                <MovieDetail
                                movie={undefined}
                />
</styled(Component)>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <MovieList
                  movies={
                                    Array [
                                                      Object {
                                                        "director": "Foo Bar",
                                                        "episode_id": 22,
                                                        "id": 1,
                                                        "opening_crawl": "foo foo foo",
                                                        "release_date": "2001-01-01",
                                                        "title": "Foo",
                                                      },
                                                      Object {
                                                        "director": "Bar Baz",
                                                        "episode_id": 13,
                                                        "id": 2,
                                                        "opening_crawl": "bar bar bar",
                                                        "release_date": "2001-02-02",
                                                        "title": "Bar",
                                                      },
                                                      Object {
                                                        "director": "Baz Biz",
                                                        "episode_id": 37,
                                                        "id": 3,
                                                        "opening_crawl": "baz baz baz",
                                                        "release_date": "2003-03-03",
                                                        "title": "Baz",
                                                      },
                                                    ]
                  }
                  onClick={[Function]}
/>,
                "className": "left",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "movies": Array [
                    Object {
                      "director": "Foo Bar",
                      "episode_id": 22,
                      "id": 1,
                      "opening_crawl": "foo foo foo",
                      "release_date": "2001-01-01",
                      "title": "Foo",
                    },
                    Object {
                      "director": "Bar Baz",
                      "episode_id": 13,
                      "id": 2,
                      "opening_crawl": "bar bar bar",
                      "release_date": "2001-02-02",
                      "title": "Bar",
                    },
                    Object {
                      "director": "Baz Biz",
                      "episode_id": 37,
                      "id": 3,
                      "opening_crawl": "baz baz baz",
                      "release_date": "2003-03-03",
                      "title": "Baz",
                    },
                  ],
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <MovieDetail
                  movie={undefined}
/>,
                "className": "right",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "movie": undefined,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
